vectorstores/alloydb/distance_strategy.go

```diff
@@ -0,0 +1,125 @@
+package alloydb
+
+import (
+	"errors"
+	"fmt"
+)
+
+// defaultDistanceStrategy is the default strategy used if none is provided
+var defaultDistanceStrategy = CosineDistance{}
+
+type distanceStrategy interface {
+	String() string
+	operator() string
+	searchFunction() string
+	similaritySearchFunction() string
+}
+
+type Euclidean struct{}
+
+func (e Euclidean) String() string {
+	return "euclidean"
+}
+func (e Euclidean) operator() string {
+	return "<->"
+}
+func (e Euclidean) searchFunction() string {
+	return "vector_l2_ops"
+}
+
+func (e Euclidean) similaritySearchFunction() string {
+	return "l2_distance"
+}
+
+type CosineDistance struct{}
+
+func (c CosineDistance) String() string {
+	return "cosineDistance"
+}
+func (c CosineDistance) operator() string {
+	return "<=>"
+}
+func (c CosineDistance) searchFunction() string {
+	return "vector_cosine_ops"
+}
+
+func (c CosineDistance) similaritySearchFunction() string {
+	return "cosine_distance"
+}
+
+type InnerProduct struct{}
+
+func (i InnerProduct) String() string {
+	return "innerProduct"
+}
+
+func (i InnerProduct) operator() string {
+	return "<#>"
+}
+
+func (i InnerProduct) searchFunction() string {
+	return "vector_ip_ops"
+}
+
+func (i InnerProduct) similaritySearchFunction() string {
+	return "inner_product"
+}
+
+// hnswOptions holds the configuration for the hnsw index.
+type hnswOptions struct {
+	m              int
+	efConstruction int
+}
+
+// ivfflatOptions holds the configuration for the ivfflat index.
+type ivfflatOptions struct {
+	lists int
+}
+
+// ivfOptions holds the configuration for the ivf index.
+type ivfOptions struct {
+	lists     int
+	quantizer string
+}
+
+// scannOptions holds the configuration for the ScaNN index.
+type scannOptions struct {
+	numLeaves int
+	quantizer string
+}
+
+// indexOptions returns the specific options for the index based on the index type
+func (index *BaseIndex) indexOptions() (string, error) {
```

Comment:
This looks better, but I think there are a few things we can improve:

1. Implement an "Index" interface:
```
type Index interface {
    Options() string
}
```

2. Each index has it's own implementation of it:
```
// ScaNNOptions holds the configuration for the ScaNN index.
type ScaNNOptions struct {
	numLeaves int
	quantizer string
}

func (s ScaNNIndex) Options() string {
    return fmt.Sprintf("(num_leaves = %d, quantizer = %s)", opts.numLeaves, opts.quantizer)
} 

```